### PR TITLE
[POC] Make [g]evented/cron services more flexible

### PR DESCRIPTION
**Pull request by Julien00859**
Task: 2146565

### CODE REVIEW

**Comment by Julien00859** [comment]
The `diff -y` comparison of `odoo-bin --help`

https://gist.githubusercontent.com/Julien00859/77d54874c712d39ba5d170522e09b183/raw/3749475fafedcaed6b999e1d5107294e9c210e19/odoo-bin-help.diff

File: odoo/tools/config.py [hunk 1]
```diff
@@ -631,13 +792,28 @@ def save(self):
...
             if opt in ('log_level',):
                 p.set('options', opt, loglevelnames.get(self.options[opt], self.options[opt]))
             elif opt == 'log_handler':
                 p.set('options', opt, ','.join(_deduplicate_loggers(self.options[opt])))
             else:
                 p.set('options', opt, self.options[opt])
 
+        p.add_section('http')
+        for opt, val in sorted(self.options_http.items(), key=operator.itemgetter(0)):
+            print('!!!')
```
Comment (Julien00859) [inline odoo/tools/config.py:806]: todo remove

**Comment by Julien00859** [comment]
@Elkasitu @xmo-odoo @odony

**Comment by xmo-odoo** [comment]
This doesn't do anything unless all the code which checks for the various bits is also modified to access the sub-variables right? For instance the HTTP workers would have to access tools.options.options_http and fallback on tools.options. And if multiple "contexts" use the same codepaths for checking, it won't necessarily work correctly. 

I was looking at it yesterday, and was thinking of a more implicit direction using configparser's support for multiple sources (or possibly chainmap) & sections: 

* set `"options"` as the default section
* replace tools.config by a proxy mapping which checks for some sort of tag (a threadlocal maybe?), then delegates to a chainmap of `[cli_options, config[tag], cli_defaults]`
* provide an API for contexts to easily set (either permanently or on a temporary basis) the current tag, this would be set by the worker threads & processes (not sure whether the "manager" process in worker mode should work with default or a specific config)

Ideally this would lead to cleaning up the configmanager some (e.g. removing my_defaults, maybe diffing the defaults dict and the parsed values instead) though there are a few complexities involved e.g. type conversion for values from the configfile, plus the subtetly of different options having different "override regimes" e.g. options which always override the config file versus options which override the config file IFF not None (though I guess we could just pop the former from the `cli_options` mapping?)

**Comment by Julien00859** [comment]
> Unless I missed something, this doesn't do anything unless all the code which checks for the various bits is also modified to access the sub-variables right?

You're right, it doesn't do anything *right now*, we still need to update the code that uses the config but let's agree on the API first :)

> For instance the HTTP workers would have to access tools.options.options_http and fallback on tools.options.

`tools.options.options_http` exists solely so we can replace the http dedicated option in their own section in the ini file. `tools.options` exposes the right value already. https://github.com/odoo/odoo/pull/50339/files#diff-3813bc998e290bd756b27aef89662e82R600-R627

> And if multiple "contexts" use the same codepaths for checking, it won't necessarily work correctly?

Gné ?

> I was looking at it yesterday, and was thinking of a more implicit direction using chainmap

Would love to ! :100: 

> Default section support

The downside of the default section is when accessing the value from the configuration, if it is not set in the dedicated section it will fallback silently in the default one. This mean the value set in the default section is priority to the value set in the command line.

i.e.

1) cli dedicated
2) config
3) cli generic

I decided *not to* use the default section and to do the fallback myself to implement the following

1) cli dedicated
2) config dedicated
3) cli generic
4) config generic

**Comment by xmo-odoo** [comment]
> > For instance the HTTP workers would have to access tools.options.options_http and fallback on tools.options.
> 
> `tools.options.options_http` exists solely so we can replace the http dedicated option in their own section in the ini file. `tools.options` exposes the right value already. https://github.com/odoo/odoo/pull/50339/files#diff-3813bc998e290bd756b27aef89662e82R600-R627

So `tools.options['whatever_http']`? Same difference.

> > And if multiple "contexts" use the same codepaths for checking, it won't necessarily work correctly?
> 
> Gné ?

HTTP and cron both call a method (possibly from the ORM), method wants to check for CPU limit, method either checks the wrong key or has to jump through hoops to try and get the right key depending on execution context.

> > Default section support
> 
> The downside of the default section is when accessing the value from the configuration, if it is not set in the dedicated section it will fallback silently in the default one. This mean the value set in the default section is priority to the value set in the command line.
> 
> i.e.
> 
>     1. cli dedicated
>     2. config
>     3. cli generic
> 
> I decided _not to_ use the default section and to do the fallback myself to implement the following
> 
>     1. cli dedicated
>     2. config dedicated
>     3. cli generic
>     4. config generic

Having the config now override the CLI seems not super useful and error prone.

**Comment by Julien00859** [comment]
https://github.com/odoo/odoo/pull/55742

**Pull request closed without merging**